staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/metrics/metrics_test.go

```diff
@@ -0,0 +1,174 @@
+limitations under the License.
+*/
+
+package metrics_test
+
+import (
+	"context"
+	"encoding/base64"
+	"fmt"
+	"strconv"
+	"strings"
+	"testing"
+
+	"k8s.io/apiserver/pkg/storage/value"
+	aestransformer "k8s.io/apiserver/pkg/storage/value/encrypt/aes"
+	kmstypes "k8s.io/apiserver/pkg/storage/value/encrypt/envelope/kmsv2"
+	"k8s.io/apiserver/pkg/storage/value/encrypt/envelope/metrics"
+	"k8s.io/component-base/metrics/legacyregistry"
+	"k8s.io/component-base/metrics/testutil"
+	kmsservice "k8s.io/kms/service"
+)
+
+const (
+	testText              = "abcdefghijklmnopqrstuvwxyz"
+	testContextText       = "0123456789"
+	testEnvelopeCacheSize = 10
+	testKeyVersion        = "1"
+	testKeyHash           = "sha256:6b86b273ff34fce19d6b804eff5a3f5747ada4eaa22f1d49c01e52ddb7875b4b"
+	testEmptyKeyHash      = "sha256:e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855"
+)
+
+// testEnvelopeService is a mock Envelope service which can be used to simulate remote Envelope services
+// for testing of Envelope based encryption providers.
+type testEnvelopeService struct {
+	annotations map[string][]byte
+	disabled    bool
+	keyVersion  string
+}
+
+func (t *testEnvelopeService) Decrypt(ctx context.Context, uid string, req *kmsservice.DecryptRequest) ([]byte, error) {
+	if t.disabled {
+		return nil, fmt.Errorf("Envelope service was disabled")
+	}
+	if len(uid) == 0 {
+		return nil, fmt.Errorf("uid is required")
+	}
+	if len(req.KeyID) == 0 {
+		return nil, fmt.Errorf("keyID is required")
+	}
+	return base64.StdEncoding.DecodeString(string(req.Ciphertext))
+}
+
+func (t *testEnvelopeService) Encrypt(ctx context.Context, uid string, data []byte) (*kmsservice.EncryptResponse, error) {
+	if t.disabled {
+		return nil, fmt.Errorf("Envelope service was disabled")
+	}
+	if len(uid) == 0 {
+		return nil, fmt.Errorf("uid is required")
+	}
+	annotations := make(map[string][]byte)
+	if t.annotations != nil {
+		for k, v := range t.annotations {
+			annotations[k] = v
+		}
+	} else {
+		annotations["local-kek.kms.kubernetes.io"] = []byte("encrypted-local-kek")
+	}
+	return &kmsservice.EncryptResponse{Ciphertext: []byte(base64.StdEncoding.EncodeToString(data)), KeyID: t.keyVersion, Annotations: annotations}, nil
+}
+
+func (t *testEnvelopeService) Status(ctx context.Context) (*kmsservice.StatusResponse, error) {
+	if t.disabled {
+		return nil, fmt.Errorf("Envelope service was disabled")
+	}
+	return &kmsservice.StatusResponse{KeyID: t.keyVersion}, nil
+}
+
+func (t *testEnvelopeService) SetDisabledStatus(status bool) {
+	t.disabled = status
+}
+
+func (t *testEnvelopeService) SetAnnotations(annotations map[string][]byte) {
+	t.annotations = annotations
+}
+
+func (t *testEnvelopeService) Rotate() {
+	i, _ := strconv.Atoi(t.keyVersion)
+	t.keyVersion = strconv.FormatInt(int64(i+1), 10)
+}
+
+func newTestEnvelopeService() *testEnvelopeService {
+	return &testEnvelopeService{
+		keyVersion: testKeyVersion,
+	}
+}
+
+func TestEnvelopeMetrics(t *testing.T) {
```

Comment:
I think this test should be part of `staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/kmsv2/envelope_test.go`. 

The tests in this package should only test the methods exposed for recording the metrics and checking they work as expected.

Reply:
I could move this whole thing to kmsv2/envelope_test.go but this test seems metrics specific and `staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/metrics/metrics.go` isn't just for kmv2. In theory, metrics_test.go should be for both kmsv1 and kmsv2. However currently there is no metrics test for kmsv1.  I'm okay either way. WDYT @enj ?